src/rust/lib/graph-editor/src/component/node/port.rs

```diff
@@ -0,0 +1,417 @@
+}
+
+/// Port definition.
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct Port<T:ShapeViewDefinition+Clone> {
+    pub data : Rc<PortData<T>>
+}
+
+/// Type that represents an input port.
+pub type InputPort = Port<InputPortView>;
+
+/// Type that represents an output port.
+pub type OutputPort = Port<OutputPortView>;
+
+/// Internal data of `Port.
+#[derive(Debug,Clone)]
+#[allow(missing_docs)]
+pub struct PortData<T:ShapeViewDefinition+Clone> {
+        spec : RefCell<Specification>,
+    pub view : Rc<component::ShapeView<T>>
+}
+
+impl<T:ShapeViewDefinition<Shape=shape::Shape>+Clone> Port<T> {
+
+    /// Constructor.
+    pub fn new(spec:Specification) -> Self {
+        let logger = Logger::new("node");
+        let view   = Rc::new(component::ShapeView::new(&logger));
+        let spec = RefCell::new(spec);
+        let data = PortData{spec,view};
+        Self{data: Rc::new(data)}.init()
+    }
+
+    fn init(self) -> Self {
+        self.update();
+        self
+    }
+
+    /// Modifies the port specification.
+    pub fn mod_specification<F:FnOnce(&mut Specification)>(&mut self, f:F) {
+        f(&mut self.data.spec.borrow_mut());
+        self.update()
+    }
+
+    /// Update the shape parameters and sprite location with values from our current specification.
+    fn update(&self) {
+        self.update_shape();
+        self.update_sprite();
+    }
+
+    /// Update the position of the sprite according to the Port specification.
+    /// The position is given along a circle, thus the position and rotation of the sprite
+    /// are tied together, so the Port always point in the right direction.
+    fn update_sprite(&self) {
+        let spec = self.data.spec.borrow();
+        let translation_vector = nalgebra::Vector3::new(0.0,spec.inner_radius,0.0);
+        let rotation_vector    = -nalgebra::Vector3::new(0.0,0.0,spec.location.rad().value);
+        let rotation           = nalgebra::Rotation3::new(rotation_vector);
+        let translation        = rotation * translation_vector;
+
+        let node = &self.data.view.display_object;
+        node.set_position(translation);
+        node.set_rotation(rotation_vector);
+    }
+
+    /// Update the shape parameters with values from our `Specification`.
+    fn update_shape(&self){
+        if let Some(t) = self.data.view.data.borrow().as_ref() {
+            let spec = self.data.spec.borrow();
+            t.shape.update_from_spec(&spec);
+        }
+    }
+}
+
+impl<'t,T:ShapeViewDefinition+Clone> From<&'t Port<T>> for &'t display::object::Node {
+    fn from(t:&'t Port<T>) -> Self {
+        &t.data.view.display_object
+    }
+}
+
+
+
+// ====================
+// === Port Manager ===
+// ====================
+
+/// Handles creation and layouting of ports around a node.
+/// TODO implement the layouting
+#[derive(Debug,Default)]
+pub struct PortManager {
+    /// The node that all ports will be placed around.
+    parent_node  : RefCell<Option<WeakNode>>,
```

Comment:
This is very bad because:

1. It is not port's responsibility to know it's parent.
2. This creates a cycle, of course it is using `WeakNode` so the cycle is broken, but any cycle is ugly.

I cannot find in code what this is used for. What do you need it for?

Reply:
This is refactored now and the parent is only needed to create a new port. 
There it is used to set the new ports parent.